src/train.py

```diff
@@ -0,0 +1,262 @@
+            
+            if terminal or ep_steps == args.max_ep_steps:  
+                # Collect total reward of episode              
+                episode_rewards.append(episode_reward)
+                # Reset episode reward and episode steps counters
+                episode_reward = 0
+                ep_steps = 0
+                # Reset environment and state buffer for next episode
+                reset_env_and_state_buffer(env, state_buf, args)          #utilsssss   --------------------   
+        
+        ## Training step    
+        # Get minibatch from replay mem
+        states_batch, actions_batch, rewards_batch, next_states_batch, terminals_batch = replay_mem.getMinibatch()
+        # Calculate target by passing next states through the target network and finding max future Q
+        #future_Q = sess.run(DQN_target.output, {state_ph:next_states_batch})
+        max_future_Q = np.max(future_Q, axis=1) #actually this one i don't know if should be utils or not ------------------
+        # Q values of the terminal states is 0 by definition
+        max_future_Q[terminals_batch] = 0
+        targets = rewards_batch + (max_future_Q*args.discount_rate)
+        
+        # Execute training step
+        if train_step % args.save_log_step == 0:
+            # Train and save logs
+            average_reward = sum(episode_rewards)/len(episode_rewards)
+            summary_str, _ = sess.run([summary_op, DQN_train_step_op], {state_ph:states_batch, action_ph:actions_batch, target_ph:targets, reward_var: average_reward, epsilon_var: epsilon})
+            summary_writer.add_summary(summary_str, train_step)
+            # Reset rewards buffer
+            episode_rewards = []
+        else:
+            # Just train
+            _ = sess.run(DQN_train_step_op, {state_ph:states_batch, action_ph:actions_batch, target_ph:targets})
+        
+        # Update target networks    
+        if train_step % args.update_target_step == 0:
+            sess.run(update_target_op) # i'm not sure where this comes from--------------------------
+        
+        # Calculate time per step and display progress to console   
+        duration = time.time() - start_time
+        duration_values.append(duration)
+        ave_duration = sum(duration_values)/float(len(duration_values))
+        
+        #sys.stdout.write('\x1b[2K\rStep {:d}/{:d} \t ({:.3f} s/step)'.format(train_step, args.num_steps_train, ave_duration))
+        #sys.stdout.flush()       
+        
+        # Save checkpoint       
+        if train_step % args.save_ckpt_step == 0:
+            saver.save(sess, checkpoint_path, global_step=train_step)
+            sys.stdout.write('\n Checkpoint saved\n')   
+            sys.stdout.flush() 
+            
+            # Reset time calculation
+            duration_values = []
+            
+                  
+
+if  __name__ == '__main__':
+    train_args = get_train_args()
+    train(train_args)         
```

Comment:
This might want to be changed (rather than command line arguments, this could be a Trainer class with all the above methods and a constructor that takes in the relevant arguments). The only reason I would recommend this is that it would make it easier to test it.